package collection.set;

import collection.set.member.Member;

public class JavaHashCodeMain {
    public static void main(String[] args) {

        //Object의 기본 hashCode는 객체의 참조값을 기반으로 생성
        Object obj1 = new Object();
        Object obj2 = new Object();
        System.out.println("obj1.hashCode() = " + obj1.hashCode());         // obj1.hashCode() = 258952499
        System.out.println("obj2.hashCode() = " + obj2.hashCode());         // obj2.hashCode() = 1067040082

        //각 클래스마다 hashCode를 이미 오버라이딩 해두었다.
        Integer i = 10;
        String strA = "A";
        String strAB = "AB";

        System.out.println("i.hashCode() = " + i.hashCode());                    // i.hashCode() = 10
        System.out.println("strA.hashCode() = " + strA.hashCode());              // strA.hashCode() = 65
        System.out.println("strAB = " + strAB);                                  // strAB = AB

        //hashCode는 마이너스 값이 들어올 수 있다.
        System.out.println("Integer.valueOf(-1).hashCode() = " + Integer.valueOf(-1).hashCode());       // Integer.valueOf(-1).hashCode() = -1

        //둘은 같을까 다를까?, 인스턴스는 다르지만, equals는 같다.
        Member member1 = new Member("idA");
        Member member2 = new Member("idA");

        //equals, hashCode를 오버라이딩 하지 않은 경우와, 한 경우를 비교
        System.out.println("(member1 == member2) = " + (member1 == member2));                   // (member1 == member2) = false     같은 객체인지
        System.out.println("(member1 equals member2) = " + (member1.equals(member2)));          // (member1 equals member2) = true  문자열이 같은지
        System.out.println("member1.hashCode() = " + member1.hashCode());                       // member1.hashCode() = 104070
        System.out.println("member2.hashCode() = " + member2.hashCode());                       // member2.hashCode() = 104070
                                                                                                // 없으면 메모리 주소로 계산해서 다르게 나오고, 직접 오버라이딩하면 내부 데이터(내용물)로 계산해서 같게 나옵니다.
    }
}
